arviz/stats/diagnostics.py

```diff
@@ -47,130 +106,135 @@ def effective_sample_size(data, *, var_names=None):
-        max_t = t + 2
-        t += 2
+    Returns
+    -------
+    xarray.Dataset
+        Return the effective sample size for mean, :math:`\hat{N}_{eff}`
 
-    # Geyer's initial monotone sequence
-    t = 3
-    while t <= max_t - 2:
-        if (rho_hat_t[t + 1] + rho_hat_t[t + 2]) > (rho_hat_t[t - 1] + rho_hat_t[t]):
-            rho_hat_t[t + 1] = (rho_hat_t[t - 1] + rho_hat_t[t]) / 2.0
-            rho_hat_t[t + 2] = rho_hat_t[t + 1]
-        t += 2
-    ess = (
-        int((n_chain * n_draws) / (-1.0 + 2.0 * np.sum(rho_hat_t)))
-        if not np.any(np.isnan(rho_hat_t))
-        else np.nan
-    )
-    return ess
+    Notes
+    -----
+    The basic ess diagnostic is computed by:
 
+    .. math:: \hat{N}_{eff} = \frac{MN}{\hat{\tau}}
+    .. math:: \hat{\tau} = -1 + 2 \sum_{t'=0}^K \hat{P}_t'
 
-def autocorr(x):
-    """Compute autocorrelation using FFT for every lag for the input array.
+    where :math:`\hat{\rho}_t` is the estimated _autocorrelation at lag t, and T
+    is the first odd positive integer for which the sum
+    :math:`\hat{\rho}_{T+1} + \hat{\rho}_{T+1}` is negative.
 
-    See https://en.wikipedia.org/wiki/autocorrelation#Efficient_computation
+    The current implementation is similar to Stan, which uses Geyer's initial monotone sequence
+    criterion (Geyer, 1992; Geyer, 2011).
 
-    Parameters
+    References
     ----------
-    x : Numpy array
-        An array containing MCMC samples
-
-    Returns
-    -------
-    acorr: Numpy array same size as the input array
+    Vehtari et al. (2019) see https://arxiv.org/abs/1903.08008
+    https://mc-stan.org/docs/2_18/reference-manual/effective-sample-size-section.html Section 15.4.2
+    Gelman et al. BDA (2014) Formula 11.8
     """
-    y = x - x.mean()
-    len_y = len(y)
-    result = fftconvolve(y, y[::-1])
-    acorr = result[len(result) // 2 :]
-    acorr /= np.arange(len_y, 0, -1)
-    acorr /= acorr[0]
-    return acorr
+    methods = {
+        "bulk": _ess_bulk,
+        "tail": _ess_tail,
+        "quantile": _ess_quantile,
+        "mean": _ess_mean,
+        "sd": _ess_sd,
+        "median": _ess_median,
+        "mad": _ess_mad,
+        "z_scale": _ess_z_scale,
+        "folded": _ess_folded,
+        "split": _ess_split,
+        "identity": _ess_identity,
+    }
+
+    methods_relative = {
+        "bulk": _ress_bulk,
+        "tail": _ress_tail,
+        "quantile": _ress_quantile,
+        "mean": _ress_mean,
+        "sd": _ress_sd,
+        "median": _ress_median,
+        "mad": _ress_mad,
+        "z_scale": _ress_z_scale,
+        "folded": _ress_folded,
+        "split": _ress_split,
+        "identity": _ress_identity,
+    }
+
+    if method not in methods:
+        raise TypeError(
+            "ESS method {} not found. Valid methods are:\n{}".format(method, "\n    ".join(methods))
+        )
+    if relative:
+        ess_func = methods_relative[method]
+    else:
+        ess_func = methods[method]
```

Comment:
I feel like I am missing something really obvious. Why are different methods defined depending on ess or ress? Would it be possible to first calculate the ess and then divide by n_draws*n_chains if `relative=True` or do nothing otherwise?

Reply:
Yes, that could be done. (etc different `ess` functions could take `relative` keyword).

In this case I wanted to keep these separate, so each function is explicit and also the `ess` functions would be as clean as possible.

Reply:
And `ess` function can also eat InferenceData or other objects, so handling relativity is easier with a func.